### ISSUE TITLE
maxParallel stalls when items are processed out of order

### ISSUE THREAD

**Issue by devinivy** [signicode/scramjet#25]
Say I would like to process some SQL tables two at a time.  Each table is mapped `into()` a stream, and each of those streams will take variable time to complete.  If table B ends before table A, then table C wont start processing until A also completes.  Here's some code!

```js
// A and B immediately start streaming.
// If B finishes streaming first, I would expect C to start streaming.
// Instead C doesn't start streaming until A also completes.

await DataStream.from(['A', 'B', 'C'])
    .setOptions({ maxParallel: 2 })
    .into(
        async (out, tableName) => await out.pull(streamTable(tableName)),
        new DataStream()
    )
    .run();
```

**Comment by MichalCz** [comment]
Hi @devinivy,

I moved this issue here. I'm making a number of small changes for a new minor release (that I promised won't happen before v5 ;) ).

So let's make it clear what we need here:

> Streams A and B and C are piped one after another to a single output, but until A or B is drained we don't want to start pulling (for instance: requesting an url) C. 
> Output order has no value at all and items are streamed from A and B in parallel.
> If B runs dry before A, C starts flowing.

The output could be:

```
a1
b1
b2
b3
a2
b4 <--- B end
a3
c1
c2
a4 <--- A end
d1
c3
...
```

Is that your expected outcome?

**Comment by devinivy** [comment]
That sounds and looks right to me!

**Comment by MichalCz** [comment]
Ok. This should be achievable using the MultiStream class, but let me think about this a little, as it may be sensible to add a new feature here.

**Comment by MichalCz** [comment]
Ok. I need to polish my code a little before I post it, but what you want to do can be achieved now with the following code (pseudo):

```javascript
stream.use(s => {
  const out = new MultiStream();
  s.each(item => {
    const str = mkStreamFrom(item);
    out.add(str);
    return str.whenEnd(); // this keeps the backpressure
  });
  return out.mux();
})
```

**Comment by devinivy** [comment]
Thank you for putting in the thought and effort!  I will need to process that a little but, but I think I see the core of the idea.  Do you have any insights as to where to `maxParallel` constant would come into play?  Is it an option set on `stream` (per your example)?

**Comment by MichalCz** [comment]
Exactly there. :)

**Comment by MichalCz** [comment]
@devinivy did you perhaps have any success in running this? I'm going to test this as a principle and put some name to such a method.

**Comment by devinivy** [comment]
Hey, I tried this out and didn't get the results I would have expected.  It seems to behave identically, except the stream may end before all items (in my original example, "tables") are processed.  Here's a script I'm using to test this out, visualizing stream progress using progress bars.  You'll see at the top of the file a constant to switch `maxParallel` behavior based upon your snippet.

https://github.com/devinivy/scramjet-experiments/blob/master/max-parallel.js

I can imagine this being a common "issue" when muxing, since it's hard to know when one is done adding streams to a `MultiStream`, and if the sub-streams complete before another one is added, then the muxed stream may end "prematurely."  I put this all in quotes because I acknowledge it's the intended behavior!  Even so, I didn't see the parallel behavior that I expected.  I will let you know if I come-up with a solution!

**Comment by MichalCz** [comment]
Hi @devinivy,

I've got a solution in a form of a new method: `unorder(func)` which does execute consecutive calls without order while keeping a set number of simultaneous operations running. It'll be released in the next minor within a couple days.

See the change set here: [scramjet-experiments devinivy:master...MichalCz:master](https://github.com/devinivy/scramjet-experiments/compare/master...MichalCz:master)

**Comment by MichalCz** [comment]
@devinivy - correction. v4.20 is out now, please check it out.

**Comment by devinivy** [comment]
@MichalCz sorry for the delay!  Thanks so much for this—it seems to work well.  A final curiosity—what kind of error propagation can one expect from exceptions throw inside `unorder()`'s callback?

**Comment by devinivy** [comment]
By the way regarding this,
> Your streams did have a minor problem of not ever being finished, because you used the readPromise but everytime it was called it received an empty array, so it kept the stream opened. 

This is a little confusing to me.  The streams did seem to finish for me once an empty array was returned.  My impression is that that's how to signal the end of the stream when using `promiseRead()`, per this code in scramjet-core https://github.com/signicode/scramjet-core/blob/master/lib/util/mk-read.js#L45-L47.  If that's not correct, what is the best way to signal the end of the stream when using `promiseRead()`?

**Comment by MichalCz** [comment]
So as to errors - the resulting stream will get a normal error which you can `.catch` and resolve or handle in `.on('error')`. I would suggest the first option though as that allows you to resolve the error with some data and reject based on your logic.

As to `promiseRead()` see that there's `done && !chunks.length` so if there's nothing to push it will end, but only if we're already done.

Now there's a need for better docs for the base class, but since it will get a major rewrite in v5 (you can check out the `series/5` branch) I wouldn't want to update the docs now, since the `PromiseTransformSteram` will be the only export of another package (`scramjet-core`) and all the current streams will migrate to this one.

**Comment by MichalCz** [comment]
Actually - I'm getting confused with this also. It seems that this is an empty array indeed. Maybe I should document this after all... :/

**Comment by devinivy** [comment]
If you'd like we can create an issue to document and/or alter this behavior of `promiseRead()`, but I think the original issue resolved—thanks! 👍

**Comment by MichalCz** [comment]
I'll get on this anyway. Version 5 is round the corner and brings some major changes. We'll still support the v4 interface (via adapter) and will add some documentation on this too.
